tests/test_hierarchy.cc

```diff
@@ -0,0 +1,221 @@
+    for (auto cell :
+         dealii::filter_iterators(dof_handler.active_cell_iterators(),
+                                  dealii::IteratorFilters::LocallyOwnedCell()))
+    {
+      cell_matrix = 0;
+      fe_values.reinit(cell);
+
+      for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+        for (unsigned int i = 0; i < dofs_per_cell; ++i)
+          for (unsigned int j = 0; j < dofs_per_cell; ++j)
+            cell_matrix(i, j) += fe_values.shape_grad(i, q_point) *
+                                 fe_values.shape_grad(j, q_point) *
+                                 fe_values.JxW(q_point);
+
+      cell->get_dof_indices(local_dof_indices);
+      constraints.distribute_local_to_global(cell_matrix, local_dof_indices,
+                                             system_matrix);
+    }
+  }
+
+public:
+  TestMeshEvaluator(const dealii::TrilinosWrappers::SparseMatrix &matrix)
+      : _matrix(matrix)
+  {
+  }
+
+private:
+  const dealii::TrilinosWrappers::SparseMatrix &_matrix;
+};
+
+BOOST_AUTO_TEST_CASE(hierarchy_2d)
+{
+  using Vector = dealii::LinearAlgebra::distributed::Vector<double>;
+  using MeshEvaluator = mfmg::DealIIMeshEvaluator<dim, Vector>;
+  using Operator = typename MeshEvaluator::global_operator_type;
+  using Mesh = mfmg::DealIIMesh<dim>;
+
+  MPI_Comm comm = MPI_COMM_WORLD;
+
+  Source<dim> source;
+
+  const int num_refinements = 6;
+
+  Laplace<dim, Vector> laplace(comm, 1);
+  laplace.setup_system(num_refinements);
+  laplace.assemble_system(source);
+
+  auto mesh =
+      std::make_shared<Mesh>(laplace._dof_handler, laplace._constraints);
+
+  const auto &a = laplace._system_matrix;
+  Vector solution(laplace._locally_owned_dofs, comm);
+  Vector rhs(laplace._system_rhs);
+
+  const int local_size = rhs.local_size();
+
+  std::default_random_engine generator;
+  std::uniform_real_distribution<typename Vector::value_type> distribution(0.,
+                                                                           1.);
+  for (int i = 0; i < local_size; i++)
+    solution.local_element(i) = distribution(generator);
+  a.vmult(rhs, solution);
+
+  auto params = std::make_shared<boost::property_tree::ptree>();
+  params->put("eigensolver: number of eigenvectors", 1);
+  params->put("eigensolver: tolerance", 1e-14);
+  params->put<unsigned int>("agglomeration: nx", 2);
+  params->put<unsigned int>("agglomeration: ny", 2);
+  params->put<unsigned int>("agglomeration: nz", 2);
+
+  TestMeshEvaluator<dim, Vector> evaluator(a);
+  mfmg::Hierarchy<MeshEvaluator, Vector> hierarchy(comm, evaluator, *mesh,
+                                                   params);
+
+  dealii::SolverControl solver_control(laplace._dof_handler.n_dofs(),
+                                       1e-12 * rhs.l2_norm());
+  dealii::SolverCG<Vector> solver(solver_control);
+
+  solution = 0.;
+  solver.solve(laplace._system_matrix, solution, rhs, hierarchy);
+
+  if (dealii::Utilities::MPI::this_mpi_process(comm) == 0)
```

Comment:
We should rework this test to check that we are getting the right order of convergence.

Reply:
Yes!